lib/web_ui/lib/src/engine/canvaskit/image_web_codecs.dart

```diff
@@ -0,0 +1,384 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// Uses the `ImageDecoder` class supplied by the browser.
+///
+/// See also:
+///
+///  * `image_wasm_codecs.dart`, which uses codecs supplied by the CanvasKit WASM bundle.
+@JS()
+library image_web_codecs;
+
+import 'dart:async';
+import 'dart:js_util' as js_util;
+import 'dart:math' as math;
+import 'dart:typed_data';
+
+import 'package:js/js.dart';
+import 'package:ui/ui.dart' as ui;
+
+import 'canvaskit_api.dart';
+import 'image.dart';
+import 'surface.dart';
+import 'surface_factory.dart';
+
+@JS('window.ImageDecoder')
+external Object? get _imageDecoderConstructor;
+
+/// Whether the current browser supports `ImageDecoder`.
+bool browserSupportsImageDecoder =
+    _imageDecoderConstructor != null;
+
+/// Sets the value of [browserSupportsImageDecoder] to its default value.
+void debugResetBrowserSupportsImageDecoder() {
+  browserSupportsImageDecoder =
+      _imageDecoderConstructor != null;
+}
+
+/// Image decoder backed by the browser's `ImageDecoder`.
+class CkBrowserImageDecoder implements ui.Codec {
+  static Future<CkBrowserImageDecoder> create({
+    required Uint8List data,
+    required String debugSource,
+    int? targetWidth,
+    int? targetHeight,
+  }) async {
+    // ImageDecoder does not detect image type automatically. It requires us to
+    // tell it what the image type is.
+    final String? contentType = detectContentType(data);
+
+    if (contentType == null) {
+      throw ImageCodecException(
+        'Failed to detect image file format of the image data using the file header, which was:\n' +
+        data.sublist(0, math.min(10, data.length)).map((int byte) => byte.toRadixString(16).padLeft(2, '0')).join(' '),
+      );
+    }
+
+    if (!await _isImageTypeSupported(contentType)) {
```

Comment:
If you don't already have an async step to hide this under, you can just create the image decoder and get the exception there instead of adding any async delay here.

Reply:
Done.